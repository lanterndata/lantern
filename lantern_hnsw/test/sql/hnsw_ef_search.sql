------------------------------------------------------------------------------
-- Test changing lantern_hnsw.ef variable at runtime 
------------------------------------------------------------------------------

\ir utils/sift1k_array.sql

CREATE INDEX hnsw_l2_index ON sift_base1k USING lantern_hnsw (v) WITH (M=5, ef=20, ef_construction=20);
SELECT * FROM ldb_get_indexes('sift_base1k');

INSERT INTO sift_base1k (id, v) VALUES 
(1001, array_fill(1, ARRAY[128])),
(1002, array_fill(2, ARRAY[128]));

-- Validate error on invalid ef_search values
\set ON_ERROR_STOP off
SET lantern_hnsw.ef = -1;
SET lantern_hnsw.ef = 0;
SET lantern_hnsw.ef = 401;
\set ON_ERROR_STOP on

-- Repeat the same query while varying ef parameter
-- NOTE: it is not entirely known if the results of these are deterministic
SET enable_seqscan=FALSE;
SELECT v AS v1001 FROM sift_base1k WHERE id = 1001 \gset

-- Queries below have the same result
SET lantern_hnsw.ef = 1;
SELECT ROUND(l2sq_dist(v, :'v1001')::numeric, 2) FROM sift_base1k order by v <-> :'v1001' LIMIT 10;

SET lantern_hnsw.ef = 2;
SELECT ROUND(l2sq_dist(v, :'v1001')::numeric, 2) FROM sift_base1k order by v <-> :'v1001' LIMIT 10;

SET lantern_hnsw.ef = 4;
SELECT ROUND(l2sq_dist(v, :'v1001')::numeric, 2) FROM sift_base1k order by v <-> :'v1001' LIMIT 10;

SET lantern_hnsw.ef = 8;
SELECT ROUND(l2sq_dist(v, :'v1001')::numeric, 2) FROM sift_base1k order by v <-> :'v1001' LIMIT 10;

SET lantern_hnsw.ef = 16;
SELECT ROUND(l2sq_dist(v, :'v1001')::numeric, 2) FROM sift_base1k order by v <-> :'v1001' LIMIT 10;

-- Queries below have the same result, which is different from above
SET lantern_hnsw.ef = 32;
SELECT ROUND(l2sq_dist(v, :'v1001')::numeric, 2) FROM sift_base1k order by v <-> :'v1001' LIMIT 10;

SET lantern_hnsw.ef = 64;
SELECT ROUND(l2sq_dist(v, :'v1001')::numeric, 2) FROM sift_base1k order by v <-> :'v1001' LIMIT 10;

SET lantern_hnsw.ef = 128;
SELECT ROUND(l2sq_dist(v, :'v1001')::numeric, 2) FROM sift_base1k order by v <-> :'v1001' LIMIT 10;

SET lantern_hnsw.ef = 256;
SELECT ROUND(l2sq_dist(v, :'v1001')::numeric, 2) FROM sift_base1k order by v <-> :'v1001' LIMIT 10;

SET lantern_hnsw.ef = 400;
SELECT ROUND(l2sq_dist(v, :'v1001')::numeric, 2) FROM sift_base1k order by v <-> :'v1001' LIMIT 10;
