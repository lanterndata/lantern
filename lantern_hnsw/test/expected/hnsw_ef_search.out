------------------------------------------------------------------------------
-- Test changing lantern_hnsw.ef variable at runtime 
------------------------------------------------------------------------------
\ir utils/sift1k_array.sql
CREATE TABLE IF NOT EXISTS sift_base1k (
    id SERIAL,
    v REAL[]
);
COPY sift_base1k (v) FROM '/tmp/lantern/vector_datasets/sift_base1k_arrays.csv' WITH csv;
CREATE INDEX hnsw_l2_index ON sift_base1k USING lantern_hnsw (v) WITH (_experimental_index_path='/tmp/lantern/files/index-sift1k-l2.usearch');
INFO:  done init usearch index
INFO:  done loading usearch index
INFO:  done saving 1000 vectors
SELECT * FROM ldb_get_indexes('sift_base1k');
   indexname   |  size  |                                                                   indexdef                                                                   | total_index_size 
---------------+--------+----------------------------------------------------------------------------------------------------------------------------------------------+------------------
 hnsw_l2_index | 720 kB | CREATE INDEX hnsw_l2_index ON public.sift_base1k USING lantern_hnsw (v) WITH (_experimental_index_path='/tmp/lantern/files/index-sift1k-l2.usearch') | 720 kB
(1 row)

INSERT INTO sift_base1k (id, v) VALUES 
(1001, array_fill(1, ARRAY[128])),
(1002, array_fill(2, ARRAY[128]));
-- Validate error on invalid ef_search values
\set ON_ERROR_STOP off
SET lantern_hnsw.ef = -1;
ERROR:  -1 is outside the valid range for parameter "lantern_hnsw.ef" (0 .. 400)
SET lantern_hnsw.ef = 0;
SET lantern_hnsw.ef = 401;
ERROR:  401 is outside the valid range for parameter "lantern_hnsw.ef" (0 .. 400)
\set ON_ERROR_STOP on
-- Repeat the same query while varying ef parameter
-- NOTE: it is not entirely known if the results of these are deterministic
SET enable_seqscan=FALSE;
SELECT v AS v1001 FROM sift_base1k WHERE id = 1001 \gset
-- Queries below have the same result
SET lantern_hnsw.ef = 1;
SELECT ROUND(l2sq_dist(v, :'v1001')::numeric, 2) FROM sift_base1k order by v <-> :'v1001' LIMIT 10;
   round   
-----------
      0.00
    128.00
 249249.00
 249285.00
 249418.00
 249515.00
 249589.00
 249647.00
 249652.00
 249675.00
(10 rows)

SET lantern_hnsw.ef = 2;
SELECT ROUND(l2sq_dist(v, :'v1001')::numeric, 2) FROM sift_base1k order by v <-> :'v1001' LIMIT 10;
   round   
-----------
      0.00
    128.00
 249249.00
 249285.00
 249418.00
 249515.00
 249589.00
 249647.00
 249652.00
 249675.00
(10 rows)

SET lantern_hnsw.ef = 4;
SELECT ROUND(l2sq_dist(v, :'v1001')::numeric, 2) FROM sift_base1k order by v <-> :'v1001' LIMIT 10;
   round   
-----------
      0.00
    128.00
 249249.00
 249285.00
 249418.00
 249515.00
 249589.00
 249647.00
 249652.00
 249675.00
(10 rows)

SET lantern_hnsw.ef = 8;
SELECT ROUND(l2sq_dist(v, :'v1001')::numeric, 2) FROM sift_base1k order by v <-> :'v1001' LIMIT 10;
   round   
-----------
      0.00
    128.00
 249249.00
 249285.00
 249418.00
 249515.00
 249589.00
 249647.00
 249652.00
 249675.00
(10 rows)

SET lantern_hnsw.ef = 16;
SELECT ROUND(l2sq_dist(v, :'v1001')::numeric, 2) FROM sift_base1k order by v <-> :'v1001' LIMIT 10;
   round   
-----------
      0.00
    128.00
 249249.00
 249285.00
 249418.00
 249515.00
 249589.00
 249647.00
 249652.00
 249675.00
(10 rows)

-- Queries below have the same result, which is different from above
SET lantern_hnsw.ef = 32;
SELECT ROUND(l2sq_dist(v, :'v1001')::numeric, 2) FROM sift_base1k order by v <-> :'v1001' LIMIT 10;
   round   
-----------
      0.00
    128.00
 249249.00
 249285.00
 249418.00
 249457.00
 249515.00
 249589.00
 249647.00
 249652.00
(10 rows)

SET lantern_hnsw.ef = 64;
SELECT ROUND(l2sq_dist(v, :'v1001')::numeric, 2) FROM sift_base1k order by v <-> :'v1001' LIMIT 10;
   round   
-----------
      0.00
    128.00
 249249.00
 249285.00
 249418.00
 249457.00
 249515.00
 249589.00
 249647.00
 249652.00
(10 rows)

SET lantern_hnsw.ef = 128;
SELECT ROUND(l2sq_dist(v, :'v1001')::numeric, 2) FROM sift_base1k order by v <-> :'v1001' LIMIT 10;
   round   
-----------
      0.00
    128.00
 249249.00
 249285.00
 249418.00
 249457.00
 249515.00
 249589.00
 249647.00
 249652.00
(10 rows)

SET lantern_hnsw.ef = 256;
SELECT ROUND(l2sq_dist(v, :'v1001')::numeric, 2) FROM sift_base1k order by v <-> :'v1001' LIMIT 10;
   round   
-----------
      0.00
    128.00
 249249.00
 249285.00
 249418.00
 249457.00
 249515.00
 249589.00
 249647.00
 249652.00
(10 rows)

SET lantern_hnsw.ef = 400;
SELECT ROUND(l2sq_dist(v, :'v1001')::numeric, 2) FROM sift_base1k order by v <-> :'v1001' LIMIT 10;
   round   
-----------
      0.00
    128.00
 249249.00
 249285.00
 249418.00
 249457.00
 249515.00
 249589.00
 249647.00
 249652.00
(10 rows)

